src/compiler/moduleNameResolver.ts

```diff
@@ -816,6 +818,34 @@ namespace ts {
         return loadNodeModuleFromDirectory(extensions, candidate, failedLookupLocations, onlyRecordFailures, state, considerPackageJson);
     }
 
+    function parseNodeModuleFromPath(path: string): { packageDirectory: string, subModuleName: string } | undefined {
+        path = normalizePath(path);
+        const nodeModules = "/node_modules/";
+        const idx = path.indexOf(nodeModules);
+        if (idx === -1) {
+            return undefined;
+        }
+
+        const indexAfterNodeModules = idx + nodeModules.length;
+        let indexAfterPackageName = moveToNextDirectorySeparatorIfAvailable(path, indexAfterNodeModules);
+        if (path.charCodeAt(indexAfterNodeModules) === CharacterCodes.at) {
+            indexAfterPackageName = moveToNextDirectorySeparatorIfAvailable(path, indexAfterPackageName);
+        }
+        const packageDirectory = path.slice(0, indexAfterPackageName);
+        const subModuleName = removeExtensionAndIndex(path.slice(indexAfterPackageName));
+        return { packageDirectory, subModuleName };
+    }
+
+    function moveToNextDirectorySeparatorIfAvailable(path: string, pos: number): number {
+        const indexOfSlash = path.indexOf(directorySeparator, pos);
+        return indexOfSlash === -1 ? pos : indexOfSlash;
+    }
+
+    function removeExtensionAndIndex(fileName: string): string {

```

Comment:
Do we have a convention where `fileName` is a path? If it someone were to input just the last part of a path here (eg: `"index.d.ts"`) then the result we would get would be `"index"`. Is that okay?

Reply:
I'll rename the parameter to `path`.